src/eckit/maths/FloatingPointExceptions.cc

```diff
@@ -0,0 +1,353 @@
+        }
+
+        if (0 != (raised & FPE_FLTUND)) {
+            LOG("code: " STR(FPE_FLTUND));
+        }
+
+        if (0 != (raised & FPE_FLTRES)) {
+            LOG("code: " STR(FPE_FLTRES));
+        }
+
+        if (0 != (raised & FPE_FLTSUB)) {
+            LOG("code: " STR(FPE_FLTSUB));
+        }
+    }
+
+#if defined(__APPLE__) && defined(__arm64__)
+    else if (signum == SIGILL) {
+        LOG("signal: " STR(SIGILL));
+
+        // On Apple Silicon a SIGFPE may be posing as a SIGILL
+        // See:
+        //    https://developer.apple.com/forums/thread/689159?answerId=733736022
+        //    https://developer.arm.com/documentation/ddi0595/2020-12/AArch64-Registers/ESR-EL1--Exception-Syndrome-Register--EL1-
+        //    https://developer.arm.com/documentation/ddi0595/2020-12/AArch64-Registers/ESR-EL2--Exception-Syndrome-Register--EL2-
+
+        const auto esr = reinterpret_cast<ucontext_t*>(ucontext)->uc_mcontext->__es.__esr;
+        constexpr decltype(esr) fpe_mask(0b10110000000000000000000000000000);
+
+        if ((fpe_mask & esr) == fpe_mask) {
+            if (0 != (esr & 0b1)) {
+                LOG("esr: IOF");
+            }
+
+            if (0 != (esr & 0b10)) {
+                LOG("esr: DZF");
+            }
+
+            if (0 != (esr & 0b0100)) {
+                LOG("esr: OFF");
+            }
+
+            if (0 != (esr & 0b1000)) {
+                LOG("esr: UFF");
+            }
+
+            if (0 != (esr & 0b10000)) {
+                LOG("esr: IXF");
+            }
+
+            if (0 != (esr & 0b10000000)) {
+                LOG("esr: IDF");
+            }
+        }
+    }
+#endif
+    else {
+        LOG("signal: ?");
+    }
+
+    LOG("---");
+
+#undef LOG
+#undef STR
+
+    CUSTOM_SIGNAL_HANDLERS.restore_all();
+    LibEcKit::instance().abort();
+
+    // Just in case we end up here, which normally we shouldn't.
+    std::abort();
+}
+
+
+}  // namespace
+
+
+//----------------------------------------------------------------------------------------------------------------------
+
+
+void FloatingPointExceptions::enable_floating_point_exceptions(const std::string& codes) {
+    if (auto excepts = make_excepts(codes); excepts != 0) {
+        enable_floating_point_exception(excepts);
+        CUSTOM_EXCEPTS |= excepts;
+    }
+}
+
+
+void FloatingPointExceptions::disable_floating_point_exceptions(const std::string& codes) {
+    if (auto excepts = make_excepts(codes); excepts != 0) {
+        disable_floating_point_exception(excepts);
+        CUSTOM_EXCEPTS &= ~excepts;
+    }
+}
+
+
+void FloatingPointExceptions::enable_custom_signal_handlers() {
+    CUSTOM_SIGNAL_HANDLERS.set(SIGFPE);
+#if defined(__APPLE__) && defined(__arm64__)
+    CUSTOM_SIGNAL_HANDLERS.set(SIGILL);
+#endif
+}
+
+
+void FloatingPointExceptions::disable_custom_signal_handlers() {
+    CUSTOM_SIGNAL_HANDLERS.restore_all();
+}
+
+
+int FloatingPointExceptions::excepts() {
+    return CUSTOM_EXCEPTS;
```

Comment:
This would only return the FE codes that were set using these functions, but not the ones that could have been set by other calls to feenableexcept, or compiler flags, or Fortran intrinsics.

Reply:
Yes and neither does it respect the exceptions set via other means